### ISSUE TITLE
Make sure date_index of test set is separated by gap from the training set

### ISSUE THREAD

**Issue by freddyaboulton** [alteryx/evalml#3078]
Came up during the review of #3071 (comment [here](https://github.com/alteryx/evalml/pull/3071#discussion_r751813099) by @ParthivNaresh ).

We used to validate the holdout set for time series by checking if the pandas indices of the dataframes were separated by  `gap + 1` units. However, this doesn't work for all possible index types like str and the woodwork index type. We should check that the `date_index` is `gap` + 1 units away before generating features and predicting. That will always be a datetime type and so the arithmetic is well defined.

**Comment by freddyaboulton** [comment]
On second thought, I think this may be blocked until we can reliably store the date index. Reason being is that in order to check if the `date_index` is `gap` + 1 units away, we need to know the units 😅

**Comment by ParthivNaresh** [comment]
So this is blocked by https://github.com/alteryx/woodwork/issues/1155?

**Comment by freddyaboulton** [comment]
@ParthivNaresh I think so

**Comment by dsherry** [comment]
https://github.com/alteryx/woodwork/issues/1155 is complete! This is now unblocked.

Is the plan to use a data check for this validation?

**Comment by freddyaboulton** [comment]
@ParthivNaresh Thanks for re-opening! I was chatting with @fjlanasa and this is what we came up with to not prevent users from calling `predict` after a successful search because the frequency can't be inferred:

- Rather than raising a ValueError in `predict`, let's refactor that logic into a helper function that could be used before calling `predict`. `predict` should no longer raise exceptions if the data violates our constraints.
- This helper function will return "something" if the test data violates our constraints. Users can then use this "something" to display warning messages prior to calling `predict`. 
- If the training data does not have an inferable frequency, let's assume the datasets are correctly separated by the gap for now. 

I'm being purposely vague here about what the "something" should be that we return because I think the api is open to design. It may be good to look at the return type of [are_ts_parameters_valid_for_split](https://github.com/alteryx/evalml/blob/main/evalml/utils/gen_utils.py#L550) for inspiration or the [partial dependence error codes](https://github.com/alteryx/evalml/blob/main/evalml/exceptions/exceptions.py#L100) (except they would be exceptions in this case).
